server/src/main/java/org/elasticsearch/index/engine/LuceneSyntheticSourceChangesSnapshot.java

```diff
@@ -0,0 +1,249 @@
+ * is enabled on the underlying index.
+ *
+ * The {@code maxMemorySizeInBytes} parameter limits the total size of uncompressed _sources
+ * loaded into memory during batch retrieval.
+ */
+public class LuceneSyntheticSourceChangesSnapshot extends SearchBasedChangesSnapshot {
+    public static final int DEFAULT_SEARCH_BATCH_SIZE = 1024;
+    public static final long DEFAULT_MEMORY_SIZE = 4 * 1024 * 1024; // 4MB
+
+    private final long maxMemorySizeInBytes;
+    private final StoredFieldLoader storedFieldLoader;
+    private final SourceLoader sourceLoader;
+
+    private int skippedOperations;
+    private long lastSeenSeqNo;
+
+    private record SearchRecord(FieldDoc doc, boolean isTombstone, long seqNo, long primaryTerm, long version, long size) {
+        int index() {
+            return doc.shardIndex;
+        }
+
+        int docID() {
+            return doc.doc;
+        }
+
+        boolean hasRecoverySourceSize() {
+            return size != -1;
+        }
+    }
+
+    private final Deque<SearchRecord> pendingDocs = new LinkedList<>();
+    private final Deque<Translog.Operation> operationQueue = new LinkedList<>();
+
+    public LuceneSyntheticSourceChangesSnapshot(
+        MappingLookup mappingLookup,
+        Engine.Searcher engineSearcher,
+        int searchBatchSize,
+        long maxMemorySizeInBytes,
+        long fromSeqNo,
+        long toSeqNo,
+        boolean requiredFullRange,
+        boolean accessStats,
+        IndexVersion indexVersionCreated
+    ) throws IOException {
+        super(engineSearcher, searchBatchSize, fromSeqNo, toSeqNo, requiredFullRange, accessStats, indexVersionCreated);
+        assert mappingLookup.isSourceSynthetic();
+        this.maxMemorySizeInBytes = maxMemorySizeInBytes;
+        this.sourceLoader = mappingLookup.newSourceLoader(SourceFieldMetrics.NOOP);
+        Set<String> storedFields = sourceLoader.requiredStoredFields();
+        this.storedFieldLoader = StoredFieldLoader.create(mappingLookup.isSourceSynthetic() == false, storedFields);
+        this.lastSeenSeqNo = fromSeqNo - 1;
+    }
+
+    @Override
+    public int skippedOperations() {
+        return skippedOperations;
+    }
+
+    @Override
+    public Translog.Operation nextOperation() throws IOException {
+        while (true) {
+            if (operationQueue.isEmpty()) {
+                loadNextBatch();
+            }
+            if (operationQueue.isEmpty()) {
+                return null;
+            }
+            var op = operationQueue.pollFirst();
+            if (op.seqNo() == lastSeenSeqNo) {
+                skippedOperations++;
+                continue;
+            }
+            lastSeenSeqNo = op.seqNo();
+            return op;
+        }
+    }
+
+    private void loadNextBatch() throws IOException {
+        List<SearchRecord> documentsToLoad = new ArrayList<>();
+        long accumulatedSize = 0;
+        boolean reachedMemoryLimit = false;
+        while (reachedMemoryLimit == false) {
```

Comment:
nit: I think we can use `accumulatedSize <= maxMemorySizeInBytes` here without `reachedMemoryLimit`.